src/bun.js/api/bun/dns_resolver.zig

```diff
@@ -2864,13 +3112,192 @@ pub const DNSResolver = struct {
+        var buffer = bun.default_allocator.alloc(u8, slice.len + 1) catch bun.outOfMemory();
+        defer bun.default_allocator.free(buffer);
+        _ = strings.copy(buffer[0..], slice);
+        buffer[slice.len] = 0;
+
+        var addr: [16]u8 = undefined;
+
+        if (c_ares.ares_inet_pton(c_ares.AF.INET, buffer.ptr, &addr) == 1) {
+            const ip = std.mem.readInt(u32, addr[0..4], std.builtin.Endian.big);
+            c_ares.ares_set_local_ip4(channel, ip);
+            return c_ares.AF.INET;
+        }
+
+        if (c_ares.ares_inet_pton(c_ares.AF.INET6, buffer.ptr, &addr) == 1) {
+            c_ares.ares_set_local_ip6(channel, &addr);
+            return c_ares.AF.INET6;
+        }
+
+        return JSC.Error.ERR_INVALID_IP_ADDRESS.throw(globalThis, "Invalid IP address: \"{s}\"", .{slice});
+    }
+
+    fn setChannelServers(channel: *c_ares.Channel, globalThis: *JSC.JSGlobalObject, callframe: *JSC.CallFrame) bun.JSError!JSC.JSValue {
+        // It's okay to call dns.setServers with active queries, but not dns.Resolver.setServers
+        if (channel != try getChannelFromVM(globalThis) and c_ares.ares_queue_active_queries(channel) != 0) {
+            return globalThis.throwValue((JSC.SystemError{
+                .code = bun.String.static("ERR_DNS_SET_SERVERS_FAILED"),
+                .message = bun.String.createFormat("Failed to set servers: there are pending queries", .{}) catch bun.outOfMemory(),
+            }).toErrorInstance(globalThis));
+        }
+
+        const arguments = callframe.arguments();
+        if (arguments.len == 0) {
+            return globalThis.throwNotEnoughArguments("setServers", 1, 0);
+        }
+
+        const argument = arguments[0];
+        if (!argument.isArray()) {
+            return globalThis.throwInvalidArgumentType("setServers", "servers", "array");
+        }
+
+        var triplesIterator = argument.arrayIterator(globalThis);
+
+        if (triplesIterator.len == 0) {
+            const r = c_ares.ares_set_servers_ports(channel, null);
+            if (r != c_ares.ARES_SUCCESS) {
+                const err = c_ares.Error.get(r).?;
+                return globalThis.throwValue(globalThis.createErrorInstance("ares_set_servers_ports error: {s}", .{err.label()}));
+            }
+            return .undefined;
+        }
+
+        const entries = bun.default_allocator.alloc(c_ares.struct_ares_addr_port_node, triplesIterator.len) catch bun.outOfMemory();
+        defer bun.default_allocator.free(entries);
+
+        var i: u32 = 0;
+
+        while (triplesIterator.next()) |triple| : (i += 1) {
+            if (!triple.isArray()) {
+                return globalThis.throwInvalidArgumentType("setServers", "triple", "array");
+            }
+
+            const family = JSValue.getIndex(triple, globalThis, 0).toInt32();
+            const port = JSValue.getIndex(triple, globalThis, 2).toInt32();
+
+            if (family != 4 and family != 6) {
```

Comment:
Do they allow `"IPv6"` and `"IPv4"` strings?

Reply:
Our implementation of dns.(promises.)Resolver.setServers in dns.ts always passes either 4, 6 or 0 to this Zig method, and the only way the Zig method could get any other value is if the user uses the undocumented Bun.dns.newResolver function and calls setServers on the return value, but it's unlikely anyone would ever want to do that